src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+
+  tiles = ceil(imageArea / MIN_TILE_AREA);
+  tiles = AVIF_MIN(tiles, MAX_TILES);
+  tiles = AVIF_MIN(tiles, MAX_THREADS);
+
+  tilesLog2 = floor(log10(tiles) / log10(2));
+
+  // If the image's width is greater than the height, use more tile columns
+  // than tile rows to make the tile size close to a square
+    
+  if (rgb->width >= rgb->height) {
+    encoder->tileRowsLog2 = tilesLog2 / 2;
+    encoder->tileColsLog2 = tilesLog2 - encoder->tileRowsLog2;
+  } else {
+    encoder->tileColsLog2 = tilesLog2 / 2;
+    encoder->tileRowsLog2 = tilesLog2 - encoder->tileColsLog2;
+  }
+
+  // It's good to have one thread per tile.
+  encoder->maxThreads = tiles;
+}
+
+/*
+   Check the result from an Avif function to see if it's an error.
+   If so, decode the error and output it, and return true.
+   Otherwise, return false.
+*/
+static avifBool isAvifError(avifResult result, char * msg) {
+  if (result != AVIF_RESULT_OK) {
+    gd_error("avif error: %s: %s", msg, avifResultToString(result));
+    return GD_TRUE;
+  }
+
+  return GD_FALSE;
+}
+
+/* Set up an avifIO object.
+   The functions in the gdIOCtx struct may point either to a file or a memory buffer.
+   To us, that's immaterial.
+   Our task is simply to assign avifIO functions to the proper functions from gdIOCtx.
+   The destroy function needs to destroy the avifIO object and anything else it uses.
+*/
+
+//TODO: can we get sizeHint somehow?
+
+static avifIO *createAvifIOFromCtx(gdIOCtx * ctx) {
+  avifIO *io;
+
+  io = (avifIO *) gdMalloc(sizeof(avifIO));
+	if (io == NULL) {
+		return NULL;
+	}
+
+  io->read = readFromCtx;
+  io->write = NULL; // this function is currently unused; see avif.h
+  io->destroy = destroyCtxAndAvifIO;
+  io->sizeHint = 0; // sadly, we don't get this information from the gdIOCtx
+  io->persistent = GD_FALSE; // TODO: This seems like the safe thing to do, but I don't know for sure. This is less efficient, because it means AVIF will make copies of all buffers.
+  io->data = ctx;
+
+  return io;
+}
+
+
+/*
+  <readfromCtx> implements the avifIOReadFunc interface by calling the relevant functions
+  in the gdIOCtx. Our logic is inspired by avifIOMemoryReaderRead() and avifIOFileReaderRead().
+  We don't know whether we're reading from a file or from memory. We don't have to know,
+  since we rely on the helper functions in the gdIOCtx.
+  We assume we've stashed the gdIOCtx in io->data, as we do in createAvifIOFromCtx().
+
+  We ignore readFlags, just as the avifIO*ReaderRead() functions do.
+
+  If there's a problem, this returns an avifResult error.
+  If things go well, return AVIF_RESULT_OK.
+  Of course these AVIF codes shouldn't be returned by any top-level GD function.
+*/
+
+static avifResult readFromCtx(avifIO * io, uint32_t readFlags, uint64_t offset, size_t size, avifROData * out)
+{
+  void *dataBuf = NULL;
+  gdIOCtx *ctx = (gdIOCtx *) io->data;
+
+  //TODO: if we set sizeHint, this will be more efficient.
+
+  if (offset > LONG_MAX || size < 0) {
```

Comment:
`size` is a `size_t` so, by definition, can never be negative, so comparing it to `< 0` will always be false

Reply:
Removing it. What you say is true. I added this because I saw something similar [in libavif](https://github.com/AOMediaCodec/libavif/blob/be4a2139d7dfe44c77ca38e678fbdf30b2471349/src/io.c#L101) - but I suppose that was there to ensure that `size` was nonzero, not <0, which it can't be.

I figured I'd allow `size` to be 0. Maybe I shouldn't.